Enable fine-grained celery configuration

jdavcs opened the pull request:
### Updated description
The following modifications are included:
1. Celery options can be configured via the `celery_conf` config option. 
2. Routing of tasks to queues is no longer hard coded. Two default routes are set in the config schema, routing `fetch_data` and `set_job_metadata` tasks to the `galaxy.external` queue (both tasks are cancellable: see `abort_when_job_stops` in `tasks.py`). As this configuration is processed by Celery, all config options specified in Celery's configuration documentation should apply (although only a handful of essentials have been explicitly tested with Galaxy - so it is absolutely possible to misconfigure stuff).
    - When specifying a route for a task, refer to the task as `galaxy.TASK` where TASK is the name of the function in `galaxy.celery.tasks` wrapped in the `galaxy_task` decorator. If, at some point, there is a submodule under `galaxy.celery.tasks` (we don't have one at this time), simply use `galaxy.SUBMODULE.TASK`.
4. The default task queue is unchanged: everything not explicitly routed will be sent to `galaxy.internal` (set in the `galaxy.celery` module)
5. The `celery_backend` and `celery_broker` config options have been removed in favor of using Celery's configuration (e.g. to set the backend, use the `celery_conf: result_backend` setting, for broker, use `celery_conf: broker_url` - as per Celery docs)

**NOTE:** I've removed the following  comment from the code because after these changes it doesn't belong in that particular location. However, it's absolutely relevant for our future Celery development:
```
183 # Should we route tasks to queues more dynamically ?
184 # That could be one way to route tasks to the resources
185 # that they require.
186 # Unfortunately it looks like discovering new queues or
187 # joining queues by a wildcard is not considered in scope
188 # for standard celery workers.
189 # We could implement that for ourselves though.
```

### Previous description
This is a simple solution to #14634 and similar use cases. We can specify anything in `galaxy.yml`. Here's an example from Celery's docs:

```py
broker_transport_options = {
    'predefined_queues': {
        'my-q': {
            'url': 'https://ap-southeast-2.queue.amazonaws.com/123456/my-q',
            'access_key_id': 'xxx',
            'secret_access_key': 'xxx',
            'backoff_policy': {1: 10, 2: 20, 3: 40, 4: 80, 5: 320, 6: 640},
            'backoff_tasks': ['svc.tasks.tasks.task1']
        }
    },
}
```
can be specified as: 

```yaml
  celery_conf:
    broker_transport_options:
      predefined_queues:
        my-q:
          url: 'https://ap-southeast-2.queue.amazonaws.com/123456/my-q'
          access_key_id: 'xxx'
          secret_access_key: 'xxx'
          backoff_policy: 
            1: 10
            2: 20
            3: 40
            4: 80
            5: 320 
            6: 640 
          backoff_tasks: 
            - 'svc.tasks.tasks.task1'
```
The loaded configuration available at `celery_app.conf.broker_transport_options` will be the same.



I am new to Celery, and I don't have specific examples of how this may be unsafe (other than specifying an invalid configuration for Celery). However, it doesn't feel safe. One solution is to have a list of allowed Celery config options - then we can filter against this list in `galaxy.config` on config load. We can start with a short list of options and expand as needed.

TODO:
- [x] to consider: only allow setting options from a list of allowed Celery options (UPDATE: no need - see discussion)
- [x] add test(s)



## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:

1. Add to `galaxy.yml`:
```yaml
celery_conf:
  task_routes:
    'galaxy.fetch_data': 'myfoo'

enable_celery_tasks: true
```
2. Launch Galaxy and 3 workers in 4 terminals:
```sh
$ GALAXY_CONFIG_FILE=config/galaxy.yml uvicorn --app-dir lib --factory galaxy.webapps.galaxy.fast_factory:factory

$ GALAXY_CONFIG_FILE=config/galaxy.yml PYTHONPATH=lib celery --app galaxy.celery worker --concurrency 2 --loglevel INFO --pool threads --queues galaxy.internal

$ GALAXY_CONFIG_FILE=config/galaxy.yml PYTHONPATH=lib celery --app galaxy.celery worker --concurrency 2 --loglevel INFO --pool threads --queues galaxy.external

$ GALAXY_CONFIG_FILE=config/galaxy.yml PYTHONPATH=lib celery --app galaxy.celery worker --concurrency 2 --loglevel INFO --pool threads --queues myfoo
```
3. Upload any data to your instance and observe the terminal output: tasks will be sent to appropriate queues (fetch_data > myfoo, set_job_metadata > galaxy.external, everything else > galaxy.internal)


## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jdavcs committed:
Enable celery configuration via galaxy's config

jdavcs committed:
Add default task routes to config schema

jdavcs committed:
Override celery task name generation

jdavcs committed:
Do not hardcode queues assignment to tasks

jdavcs committed:
Remove unused code (exec celery module as a program)

(This seems to be a leftover from an initial WIP in #11720)

jdavcs committed:
Cleanup celery module namespace, refactor

jdavcs committed:
Remove celery_broker from config_schema, use celery_conf instead

jdavcs committed:
Use celery_conf[broker_url] instead of celery_broker

jdavcs committed:
Remove celery_backend from config_schema, use celery_conf instead

jdavcs committed:
Drop reference to old config property

jdavcs committed:
Check for celery_conf, raise appropriate error

jdavcs committed:
Use celery_conf[result_backend] instead of celery_backend

mvdbeek commented on lib/galaxy/tools/execute.py:
```diff
@@ -180,19 +180,11 @@ def execute_single_job(execution_slice, completed_job):
-                # We could implement that for ourselves though.
-                # For now we just hardcode galaxy.internal (default, with access to db etc) and galaxy.external (cancelable).
-                | fetch_data.s(job_id=job_id).set(queue="galaxy.external")
+                | fetch_data.s(job_id=job_id)
```
I think that note still applies ? We've just moved the static queue elsewhere

jdavcs replied:
It does! I just thought this might be not the best place for it since we are no longer assigning a task to a queue here (so that's not where I would look for it). I was going to include the note verbatim in the PR description so it's preserved. It's a small but useful bit of dev knowledge that's not quite enough for a whole issue, and we don't have a designated wiki or something for this.. If you think having it in the code base is better, I can leave it here or, maybe better, add it at the top of the celery module?

mvdbeek commented on lib/galaxy/jobs/runners/__init__.py:
```diff
@@ -378,6 +379,19 @@ def _walk_dataset_outputs(self, job: model.Job):
+        else:
+            result_backend = celery_conf.get("result_backend", None)
+            if result_backend != "rpc://localhost":
+                raise ConfigurationError(f"Wrong celery backend: {result_backend}")
```
I ... think that as long as there is any backend this is fine, we can drop this, that was for debugging tests.

mvdbeek commented on lib/galaxy/config/sample/galaxy.yml.sample:
```diff
@@ -2392,17 +2392,24 @@ galaxy:
+  # For details, see Celery documentation at
+  # https://docs.celeryq.dev/en/stable/userguide/configuration.html.
+  #celery_conf:
+    task_routes:
```
I'm not seeing where the internal tasks are routed ?

jdavcs replied:
We set `"galaxy.internal"` as the `task_default_queue` in the celery module - so anything not explicitly routed falls back to it. (I've also tested this extensively by launching multiple workers, each listening to its own queue: tasks are handled by appropriate workers as expected)

mvdbeek reviewed (commented):
Looking good!

jdavcs committed:
Celery config option `result_backend` is required

jdavcs committed:
Add basic celery unit tests

Rename periodic tasks to match galaxy task naming convention
(galaxy.celery.tasks.foo >> galaxy.foo)

jdavcs committed:
Fix config error in tests
